src/osyris/spatial/spatial.py

```diff
@@ -0,0 +1,147 @@
+                  ], [-np.sin(azimuth), np.cos(azimuth), 0]],
+                 dtype=object)
+    return R
+
+
+def get_ang_mom(subdomain, dr_L):
+    """
+    Compute angular momentum vector in sphere of radius dr_L
+    """
+    sphere = (subdomain["amr"]["position"].norm <= dr_L).values
+    pos = subdomain["amr"]["position"][sphere]
+    mv = subdomain["hydro"]["mass"][sphere] * subdomain["hydro"]["velocity"][sphere]
+    L = np.sum(pos.cross(mv))
+    return L
+
+
+def change_basis(subdomain, new_basis):
+    """
+    Rotates all vectors in dataset to align with vector in 'new_basis'
+    """
+    try:
+        old_basis = subdomain.basis
+    except AttributeError:
+        old_basis = [0, 0, 1]  # assume it's the ramses grid
+    if hasattr(new_basis, 'nvec'):  # if it's a vector
+        new_basis = [new_basis.x.values, new_basis.y.values, new_basis.z.values]
+    rot_angle = np.arccos(
+        np.dot(old_basis, new_basis) /
+        (np.linalg.norm(old_basis) * np.linalg.norm(new_basis)))
+    rot_vector = np.cross(new_basis, old_basis)
+    R = rotation_matrix(rot_vector, rot_angle)
+    for g in subdomain.groups.keys():
+        for element in subdomain[g]:
+            if hasattr(subdomain[g][element], 'nvec'):
+                # all of this will be simplified once matmul is integraded into Vector
+                vector = np.array([
+                    subdomain[g][element].x.values, subdomain[g][element].y.values,
+                    subdomain[g][element].z.values
+                ])
+                vector = R @ vector
+                subdomain[g][element] = subdomain[g][element].__class__(
+                    x=vector[0],
+                    y=vector[1],
+                    z=vector[2],
+                    unit=subdomain[g][element].unit)
+    subdomain.basis = new_basis
+
+
+def get_spherical_colatitude(position_vec):
+    """
+    Compute colatitude from cartesian position vector
+    """
+    X = position_vec.x.values
+    Y = position_vec.y.values
+    Z = position_vec.z.values
+
+    theta = np.arctan2(np.linalg.norm([X, Y], axis=0), Z)
+
+    return position_vec.x.__class__(values=theta, unit='rad', name="position_phi")
+
+
+def get_spherical_azimuth(position_vec):
+    """
+    Compute colatitude from cartesian position vector
+    """
+    X = position_vec.x.values
+    Y = position_vec.y.values
+    phi = np.arctan2(Y, X)
+
+    return position_vec.x.__class__(values=phi, unit='rad', name="position_phi")
+
+
+def get_spherical_components(vec, comp):
+    """
+    Get the spherical components of an arbitrary cartesian vector 'vec'
+    """
+    if not hasattr(vec, 'nvec'):
+        raise ValueError("This is not a vector")
+    if vec.parent.name == "hydro":
+        position = vec.parent.parent["amr"]["position"]
+    else:
+        position = vec.parent["position"]
+
+    theta = position.theta
+    phi = position.phi
+
+    if comp.lower() == "radius":
```

Comment:
I don't think it makes much sense to group the code for the different components into a single function, and then have branching depending on the component requested.

In addition, it's a bit weird that the implementation for the vector class in `core` depends on code that is in a `spatial` submodule.

For these two reasons, I think this implementation should be inlined into the `r, theta, phi` methods of the Vector class.

Reply:
In hindsight, this would indeed be better if inlined.